### PR TITLE
[MRG] Bures-Wasserstein Gradient Descent for Bures-Wasserstein Barycenters

### PR DESCRIPTION
**Pull request by clbonet**
## Types of changes

This PR aims to add the Bures-Wasserstein gradient descent solver to compute Bures-Wasserstein barycenters (see e.g. [Gradient descent algorithms for Bures-Wasserstein barycenters](https://proceedings.mlr.press/v125/chewi20a.html) or [Averaging on the Bures-Wasserstein manifold: dimension-free convergence of gradient descent](https://papers.neurips.cc/paper_files/paper/2021/hash/b9acb4ae6121c941324b2b1d3fac5c30-Abstract.html)).

- Restructured `ot.gaussian.bures_wasserstein_barycenter` to allow to use different methods
- Added the previous fixed-point algorithm in `ot.gaussian.bures_barycenter_fixpoint`
- Added the Bures-Wasserstein gradient descent in `ot.gaussian.bures_barycenter_gradient_descent`
- Added an iteration over the methods in the test `test_bures_wasserstein_barycenter`
- Added a test `test_fixedpoint_vs_gradientdescent_bures_wasserstein_barycenter`
- Added batch version of `ot.gaussian.bures_wasserstein_distance`
- Trace can be computed for batchs of matrices. The choices of the implementation of the trace are made following the runtimes reported by the following codes (on a CPU):

```
import torch
import jax
import numpy as np
import tensorflow as tf
import jax.numpy as jnp

A = np.random.rand(1000, 100, 100)

%timeit np.einsum("...ii", A) # 109 μs ± 1.71 μs per loop
%timeit np.trace(A, axis1=-2, axis2=-1) # 116 μs ± 1.79 μs
%timeit A.diagonal(axis1=-2, axis2=-1).sum(-1) # 114 μs ± 2.87 μs per loop

A = torch.rand(1000, 100, 100)

%timeit torch.einsum("...ii", A)  # 3.17 ms ± 1.1 ms per loop 
%timeit A.diagonal(dim1=-2, dim2=-1).sum(-1) # 3.1 ms ± 879 μs per loop

A = tf.random.uniform((1000, 100, 100))

@tf.function
def trace_sum(A):
    return tf.einsum("...ii", A)

@tf.function
def trace_sum_v2(A):
    return tf.reduce_sum(tf.linalg.diag_part(A), axis=-1)

# Warm-up execution
trace_sum(A)  
trace_sum_v2(A)

# Benchmarking
%timeit trace_sum(A) # 486 μs ± 21.1 μs per loop
%timeit trace_sum_v2(A) # 430 μs ± 36.1 μs per loop
%timeit tf.linalg.trace(A) # 404 μs ± 18.2 μs per loop

# For jax, the results might look different using jit
A = jnp.ones((1000, 100, 100)) 

%timeit jnp.einsum("...ii", A) # 13.6 ms ± 324 μs per loop
%timeit jax.vmap(jnp.trace)(A) # 12.1 ms ± 457 μs per loop
%timeit A.diagonal(axis1=-2, axis2=-1).sum(-1) # 1.64 ms ± 3.62 ms per loop
```

## Motivation and context / Related issue

The Bures-Wasserstein gradient descent comes with convergence guarantees to solve Bures-Wasserstein barycenters. Moreover, it can also be used in a stochastic way when there are too much Gaussian. Thus, it is a good alternative to the fixed-point algorithm currently implemented.

## How has this been tested (if it applies)

I added a test `test_fixedpoint_vs_gradientdescent_bures_wasserstein_barycenter` to assess both methods returns the same barycenter. I also added the itertools to `test_bures_wasserstein_barycenter`.

## PR checklist

- [x] I have read the [**CONTRIBUTING**](https://pythonot.github.io/contributing.html) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/utils.py [hunk 1]
```diff
@@ -1308,3 +1308,15 @@ def proj_SDP(S, nx=None, vmin=0.):
         Q = nx.einsum('ijk,ik->ijk', P, w)  # Q[i] = P[i] @ diag(w[i])
         # R[i] = Q[i] @ P[i].T
         return nx.einsum('ijk,ikl->ijl', Q, nx.transpose(P, (0, 2, 1)))
+
+
+def exp_bures(Sigma, S):
```
Comment (rflamary) [inline ot/utils.py:1313]: should gave an optional nx heer so that you do bnot need to do get_backend(S) every time

File: ot/gaussian.py [hunk 2]
```diff
@@ -425,15 +409,200 @@ def bures_wasserstein_barycenter(m, C, weights=None, num_iter=1000, eps=1e-7, lo
...
+            grad_bw = Id - nx.mean(nx.einsum("ij,njk,kl -> nil", Cb12_, M, Cb12_), axis=0)
+        else:  # gradient descent
+            M = nx.sqrtm(nx.einsum("ij,njk,kl -> nil", Cb12, C, Cb12))
+            grad_bw = Id - nx.sum(nx.einsum("ij,njk,kl -> nil", Cb12_, M, Cb12_) * weights[:, None, None], axis=0)
+
+        Cnew = exp_bures(Cb, - step_size * grad_bw)
+
+        # Right criteria? (for GD, seems fine, but for SGD?)
+        # check convergence
+        diff = nx.norm(Cb - Cnew)
```
Comment (rflamary) [inline ot/gaussian.py:499]: agreed not sure we want this one, maybe we should compute diff   over an epoch?

**Review by rflamary — COMMENTED** [review]
Small comments. I will let @antoinecollas do a proper review he is the expert in Riemannian optimization

File: ot/gaussian.py [hunk 3]
```diff
@@ -433,15 +416,240 @@ def bures_wasserstein_barycenter(
...
+    for it in range(num_iter):
+        Cb12 = nx.sqrtm(Cb)
+        Cb12_ = nx.inv(Cb12)
+
+        if batch_size is not None and batch_size < n:  # if stochastic gradient descent
+            if batch_size <= 0:
+                raise ValueError(
+                    "batch_size must be an integer between 0 and {}".format(n)
+                )
+            inds = np.random.choice(
```
Comment (rflamary) [inline ot/gaussian.py:504]: why is that not tested?

File: ot/gaussian.py [hunk 4]
```diff
@@ -433,15 +416,240 @@ def bures_wasserstein_barycenter(
...
+            log=log,
+            step_size=step_size,
+            batch_size=batch_size,
+        )
+    elif method == "fixed_point":
+        out = bures_barycenter_fixpoint(
+            C, weights=weights, num_iter=num_iter, eps=eps, log=log
+        )
+    else:
+        raise ValueError("Unknown method '%s'." % method)
```
Comment (rflamary) [inline ot/gaussian.py:646]: alseo test raise here

**Review by rflamary — COMMENTED** [review]
This is great. A few tests especialy about errors are missing

**Review by rflamary — COMMENTED** [review]
Just a few questions and then we can merge

File: RELEASES.md [hunk 5]
```diff
@@ -8,6 +8,10 @@
 - Automatic PR labeling and release file update check (PR #704)
 - Reorganize sub-module `ot/lp/__init__.py` into separate files (PR #714)
 - Fix documentation in the module `ot.gaussian` (PR #718)
+- Refactored `ot.bregman._convolutional` to improve readability (PR #709)
```
[the remark this answers is not in the document]
Comment (clbonet) [reply RELEASES.md:11]: Mmh, I think I did a mistake when merging with the master at some point. (It was deleted from Line 46 of the Releases.md, and it seemed to be in the wrong releases of POT)

**Review by rflamary — APPROVED** [review]
